### ISSUE TITLE
Remember previously used / discovered device addresses and retry them

### ISSUE THREAD

**Issue by calmh** [syncthing/syncthing#3474]
In addition to global and local discovery, try any addresses we've recently seen a given device use.

**Comment by AudriusButkevicius** [comment]
Also, after ClusterConfig, we can send KnownAddresses message for any peers that were part of ClusterConfig

**Comment by calmh** [comment]
Yep, exactly what I was coming here to write, as that came out of the other discussion I just didn't get it until now.

That KnownAddresses message could also be spontaneously generated along the way, that is when I connect to device A I also send a KnownAddresses update to devices B and C to let them know about the address I managed to find for A.

**Comment by AudriusButkevicius** [comment]
Yep, the question is weather we send only currently known alive addresses or even the old addresses?
Also, this could be done fairly easily with the discovery mux, yet the functionality bleeds from model back to discovery which is strange

**Comment by masgo** [comment]
Sounds good. This means if I happen to know/contact any device from a group, it will inform me about all other in the group.

**Comment by Ferroin** [comment]
The other thing to think about is if we want inform peers of our alternative addresses.  For example, if A and B are talking over IPv4 directly, should they inform each other of any IPv6 or relay addresses they have too?  This would increase the bandwidth requirements, but would probably make it even less likely that talking to the discovery server is necessary to establish a connection.

**Comment by Ferroin** [comment]
That's kind of my thought too, it's not very expensive, and it reduces the need for global discovery even further.  If implemented such that this info is only gathered when a new device connects or an existing one changes address, there should be almost no overhead compared to the rest of the protocol.
